src/main/java/com/boot/cart/service/CartService.java

```diff
@@ -1,255 +1,251 @@
         }
 
-        User user;
+        if (productDTO.getStock() == 0) {
+            throw new InvalidInputDataException("We are sorry, but currently: " + productName + " is out of order!");
+        }
+
+        if (productDTO.getStock() < quantity) {
+            throw new InvalidInputDataException("You can not remove more than: " + productDTO.getStock() + " "
+                    + productName + " Products from your shopping cart!");
+        }
+        UserDTO user;
         try {
-            user = UserMapper.DtoToUserEntity(userServiceClient.callGetUserByEmail(email));
+            user = userServiceClient.callGetUserByEmail(email);
         } catch (HttpClientErrorException.NotFound e) {
-            throw new EntityNotFoundException("UserName: " + email + " not found in the Database!");
+            throw new EntityNotFoundException("Email: " + email + " not found in the Database!");
         }
-        Cart cart = cartRepository.findByUser(user);
-        List<Product> productList = cart.getProductList();
+        Cart cart = cartRepository.findByUserId(user.getId());
 
-        Long productsInCart = productList.stream().filter(p -> p.getProductName().equals(productName)).count();
+        List<CartEntry> cartEntries;
+        long initialProductTotal = 0;
+        Integer initialQuantity = 0;
 
+        cartEntries = cart.getEntries();
+        CartEntry cartEntry = cartEntries.stream().filter(entry -> productName.equals(entry.getProductName())).findFirst().orElse(null);
 
-
-        if (productsInCart == 0) {
-            throw new InvalidInputDataException(
-                    "You currently have " + productsInCart + " Products: " + productName + " in cart!");
+        if (cartEntry == null) {
+            throw new EntityNotFoundException("CartEntry not found in the Database!");
         }
+        else{
+            initialQuantity = cartEntry.getQuantity();
+            cartEntry.setQuantity(0);
+            cartEntry.setQuantity(quantity);
+            cartEntry.setCart(cart);
 
-        double productTotalRemove = 0;
-
-        List<Product> products = productList.stream().filter(p -> p.getProductName().equals(productName)).collect(Collectors.toList());
-        productList.removeAll(products);
-
-        cart.setTotal(cart.getTotal() - (products.size() * product.getProductPrice()));
-
-        for (int i = 0; i < quantity; i++) {
-            productList.add(product);
-            productTotalRemove = product.getProductPrice() + productTotalRemove;
+            initialProductTotal = initialQuantity *  productDTO.getPrice();
         }
-        cart.setTotal(cart.getTotal() + productTotalRemove);
 
-        cart.setUser(user);
+        cart.setEntries(cartEntries);
 
-        cart.setProductList(productList);
+        cart.setTotal(cart.getTotal() - initialProductTotal);
+        cart.setTotal(cart.getTotal()+(quantity*productDTO.getPrice()));
```

Comment:
I think it could look better if you calculate cart new total in one line and set that value to the cart on the following line

Reply:
Done!